bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java

```diff
@@ -3451,6 +3447,19 @@ public void setStyle (TextStyle style, int start, int end) {
 	int length = text.length();
 	if (length == 0) return;
 	if (start > end) return;
+
+	String startString = originalText.substring(0, start+1);
+	int stringLength = startString.length();
+
+	startString = startString.charAt(stringLength-1) == '\r' ? startString.substring(0,stringLength-1) : startString;
+	int startCount = (int) startString.chars().filter(ch -> ch == '\r').count();
+
+	String endString = originalText.substring(0, end+1);
+	int endCount = (int) endString.chars().filter(ch -> ch == '\r').count();
+
+	start = start == 0 ? start : start - startCount;
+	end = end - endCount;
+
```

Comment:
I think the performance of setStyle is rather critical and this block of code looks scary both in complexity and in performance.  Several strings are created, only to be scanned with a stream and an lamda, just for counting the number of \r and very often there won't be one at all.

I think it would be far better to use java.lang.String.indexOf(int, int) to look for \r rather than to create a bunch of Strings, streams, and lambdas.

E.g., something like this **untested** code:
```
		int startCount = 0;
		int endCount = 0;
		for (int crIndex = originalText.indexOf('\r', 0); crIndex >= 0; crIndex = originalText.indexOf('\r',
				crIndex + 1)) {
			if (crIndex <= start) {
				++startCount;
			} else if (crIndex <= end) {  // Not sure if end offset is inclusive...
				++endCount;
			} else {
				break;
			}
		}

		start -= startCount;
		end -= endCount;
```

I also wonder how a \r not followed by a \n is handled overall...

Given the complexity here, a unit test would seem like a very good thing...

Reply:
As per your suggestion i have the same fix with some additional changes. I have uploaded the newer version with  java.lang.String.indexOf(int, int) being used. 
@merks : You may want to review now.

Also attached the snippet in manual tests.
Also attaching the excel sheet i have tried to test this fix(for your reference).
[Issue184_21stAug2024.xlsx](https://github.com/user-attachments/files/16687533/Issue184_21stAug2024.xlsx)